llvm/lib/MCA/InstrBuilder.cpp

```diff
@@ -89,11 +89,11 @@ static void initializeUsedResources(InstrDesc &ID,
       AllInOrderResources &= (PR.BufferSize <= 1);
     }
 
-    CycleSegment RCy(0, PRE->ReleaseAtCycle, false);
+    CycleSegment RCy(PRE->AcquireAtCycle, PRE->ReleaseAtCycle, false);
```

Comment:
This code here made me wonder if we are accounting for AcquireAtCycle correctly. 

The `NumCyclesUsed` class may only represent the total number of cycles used, without saying anything about when the resource is acquired or released. `CycleSegment` class looks like it is the one who is responsible for taking this information into account. After looking into it, I found that although this information is being tracked, it is not being used as needed:

In `ResourceManager::issueInstruction`, we say `BusyResources[Pipe] += CS.size()`, which is a map from pipe to how many cycles are left before the pipe is usable again. The problem is that the this map data structure does not model the fact that a resource can be free from issue cycle until acquire at cycle. My thought is to add another map that tracks this data, and in conjunction with `BusyResources` we can tell the full story.

The other part that needs some fixing is that `reserveResource` is called in `issueInstruction`. We need to model that a resource may not need to be reserved on instruction issue, but on acquire at cycles after instruction issue. It may be a good idea to move the call to `reserveResource` from `ResourceManager::issueInstruction` to `ResourceManager::cycleEvent`.

I will take a look into this once I get some bandwidth. I am leaving this comment here and marking this PR as needs changes in the meantime.

Reply:
> This code here made me wonder if we are accounting for AcquireAtCycle correctly.
> 
> The `NumCyclesUsed` class may only represent the total number of cycles used, without saying anything about when the resource is acquired or released. `CycleSegment` class looks like it is the one who is responsible for taking this information into account. After looking into it, I found that although this information is being tracked, it is not being used as needed

Correct.

If you see my comment from three weeks ago, this is literally what I was trying to say when I wrote "this patch doesn't add support for AcquireAtCycle".
The ResourceManager still assumes that every resource is consumed at relative cycle zero (i.e. relative to the issue event).

Reply:
I missed your comment above. I think your point is that we need to be able to do this without slowing things down too much. I have a rough idea on how me might do this in my head, but I am not sure when I will have time to pick this task back up again, since it is much less straightforward than I had for the original patch.

Reply:
> I missed your comment above. I think your point is that we need to be able to do this without slowing things down too much. I have a rough idea on how me might do this in my head, but I am not sure when I will have time to pick this task back up again, since it is much less straightforward than I had for the original patch.

No problem. It is something that I always wanted to do in my spare time but never ended up doing.
Redesigning the ResourceManager is definitely not straightforward; the simulation time is heavily dependent on how quickly it replies to individual dispatch/issue requests.
It won't be easy to find a design which doesn't hurt performance too much. In the end we will inevitably have to compromise a bit between simulation speed and simulation accuracy. That is expected.

The complexity of the new design also depends on how many pending CycleSegments can be "issued" for every resource. I am mainly thinking about out-of-order processors where multiple delayed resource consumptions can be pending at the same time. Not sure if that makes sense...